core/consensus/common/src/import_queue.rs

```diff
@@ -97,227 +96,448 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
+	}
 
-		let blocks_hashes: Vec<B::Hash> = new_blocks.1.iter().map(|b| b.hash.clone()).collect();
-		if !import_many_blocks(
-			&*block_import,
-			&link,
-			Some(&*qdata),
-			new_blocks,
-			verifier.clone(),
-		) {
-			break;
+	fn handle_is_importing(&self, hash: B::Hash, reply_sender: Sender<bool>) {
+		let _ = reply_sender.send(self.queue_blocks.contains(&hash));
+	}
+
+	fn handle_import_justification(&self, who: Origin, hash: B::Hash, number: NumberFor<B>, justification: Justification) {
+		let success = self.justification_import.as_ref().map(|justification_import| {
+			justification_import.import_justification(hash, number, justification).is_ok()
+		}).unwrap_or(false);
+		if let Some(link) = self.link.as_ref() {
+			link.justification_imported(who, &hash, number, success);
 		}
+	}
+
+	fn handle_import_blocks(&mut self, origin: BlockOrigin, blocks: Vec<IncomingBlock<B>>) {
+		trace!(target:"sync", "Scheduling {} blocks for import", blocks.len());
 
-		let mut queue_blocks = qdata.queue_blocks.write();
-		for blocks_hash in blocks_hashes {
-			queue_blocks.remove(&blocks_hash);
+		let new_best_importing_number = blocks
+			.last()
+			.and_then(|b| b.header.as_ref().map(|h| h.number().clone()))
+			.unwrap_or_else(|| Zero::zero());
+		self.queue_blocks
+			.extend(blocks.iter().map(|b| b.hash.clone()));
+		if new_best_importing_number > self.best_importing_number {
+			self.best_importing_number = new_best_importing_number;
 		}
+		self.worker_sender
+			.send(BlockImportWorkerMsg::ImportBlocks(origin, blocks))
+			.expect("1. This is holding a sender to the worker, 2. the worker should not quit while a sender is still held");
 	}
+}
 
-	trace!(target: "sync", "Stopping import thread");
+struct BlockImportWorker<B: BlockT, V: Verifier<B>> {
+	result_sender: Sender<BlockImportWorkerMsg<B>>,
+	block_import: SharedBlockImport<B>,
+	verifier: Arc<V>,
+}
+
+impl<B: BlockT, V: 'static + Verifier<B>> BlockImportWorker<B, V> {
+	pub fn new(
+		result_sender: Sender<BlockImportWorkerMsg<B>>,
+		verifier: Arc<V>,
+		block_import: SharedBlockImport<B>,
+	) -> Sender<BlockImportWorkerMsg<B>> {
+		let (sender, port) = channel::unbounded();
+		let _ = thread::Builder::new()
+			.name("ImportQueueWorker".into())
+			.spawn(move || {
+				let worker = BlockImportWorker {
+					result_sender,
+					verifier,
+					block_import,
+				};
+				for msg in port.iter() {
+					// Working until all senders have been dropped...
+					match msg {
+						BlockImportWorkerMsg::ImportBlocks(origin, blocks) => {
+							worker.import_a_batch_of_blocks(origin, blocks)
+						}
+						_ => unreachable!("Import Worker does not receive the Imported message"),
+					}
+				}
+			})
+			.expect("ImportQueueWorker thread spawning failed");
+		sender
+	}
+
+	fn import_a_batch_of_blocks(&self, origin: BlockOrigin, blocks: Vec<IncomingBlock<B>>) {
+		let count = blocks.len();
+		let mut imported = 0;
+
+		let blocks_range = match (
+			blocks
```

Comment:
This formatting is weird, it should be fine the keep the expression on one line like:
`blocks.first().and_then(|b| b.header.as_ref().map(|h| h.number())),`